clang/include/clang/Basic/DiagnosticDriverKinds.td

```diff
@@ -352,6 +352,10 @@ def warn_drv_sycl_offload_target_duplicate : Warning<
 def warn_drv_sycl_target_missing : Warning<
   "linked binaries do not contain expected '%0' target; found targets: '%1'">,
   InGroup<SyclTarget>;
+def err_drv_no_rdc_sycl_target_missing : Error<
+  "linked binaries do not contain expected '%0' target; found targets: '%1', this is not supported with '-fno-sycl-rdc'">;
+def err_drv_option_unsupported_device_code_split : Error<
```

Comment:
Do I understand correctly that `per_source` split is allowed here to preserve compatibility with existing  build scripts where it might be used already? Because other than that, it doesn't make any sense to request a per-source split if every translation unit is anyway going to be processed separately without being linked to any other translation unit.

What about `auto` device code split mode? Is it allowed in conjunction with this new compilation mode? I think it is important to allow it here too, because that is a default value of `-fsycl-device-code-split` option.

Reply:
@AlexeySachkov Thanks for the review.

1) No, it's not to preserve compatibility, it's for correctness. Let's take an example of existing behavior:

Format is: input to sycl-post-link-> sycl-post-link -> output of sycl-post-link

a) split=none

Single huge module with all device code -> sycl post link -> Single huge module with all device code

b) split=per_source  

Single huge module with all device code -> sycl post link -> One output module per source file

c) split=per_kernel

Single huge module with all device code -> sycl post link -> One output module per kernel

To implement 2 and 3, sycl-post-link copies over any functions called in the split. So if you have a per_source split, and it calls functions in another source file (which is valid because remember the input IR is all device code!), it will include that function from the other source file in the output

Now let's consider the same cases where -fno-sycl-rdc is enabled

a) split=none, -fno-sycl-rdc

We never have a single huge module, so we can't support this

b) split=per_source, -fno-sycl-rdc

For each translation unit:
   Small module with only this translation unit device code -> sycl post link -> One output per source (so always 1    because of input)


c) split=per_kernel, -fno-sycl-rdc

For each translation unit:
   Small module with only this translation unit device code -> sycl post link -> One output per kernel



So in the split=per_source, -fno-sycl-rdc or  split=per_kernel, -fno-sycl-rdc cases, if a function that is not in the source file is called, it won't be part of the input to sycl-post-link anymore because it's not one big module, so we can't include it in the split, so the output is wrong.

If this explanation was confusing please ping me.


2) About split=auto: This is a good point. Right now I error because sometimes sycl-post-link may choose split=none internally when you give it split=auto (see [here ](https://github.com/intel/llvm/blob/sycl/llvm/tools/sycl-post-link/ModuleSplitter.cpp#L85). In those cases split=none is required, we can't just say use per_source or per_kernel if -fno-sycl-rdc, so if we were to support split=auto we would have to add an error in sycl-post-link if it needs to pick split=none. But we could support auto if it internally picks split=per_source or per_kernel,

Reply:
> Why so? If the source code doesn't have any cross-translation unit calls, then it should be supported.

Because the entire point of -fno-sycl-rdc is to never create the one big module. If we wanted to support split=none, we would have to link all of the outputs of sycl-post-link back together into one big module before sending to llvm-spirv. It might not be that simple either, maybe there is other stuff you need to do for correctness. I don't see a reason why anyone would pass -fno-sycl-rdc but want split=none, so that's why there's the error.

> 
> > Do you recommend supporting split=auto?
> 
> Since this is a default mode, I think it is important from user experience point of view to support this. Why require any extra flags from user? Alternative would be to change device code split default mode based on presence of `-fno-sycl-rdc`, but I wouldn't recommend that.
> 
> What I would recommend is the following: [5.10.1. SYCL functions and member functions linkage](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:syclexternal) says that every function used from a kernel _must_ be defined in the same translation unit unless implementation supports offline linking, which is indicated by defining `SYCL_EXTERNAL` macro. I suggest that we conditionally define that macro (i.e. we don't define it in presence of `-fno-sycl-rdc`) and by doing so we will provide users with early indications that their source code is incompatible with this compilation mode. On top of that, it provides us with guarantee that at `sycl-post-link` level we see the whole call graph without undefined external functions and we should be able to safely perform any kinds of split we want to perform.

I believe this change provides the guarantee you mention but it is implemented in a different way. I added a check to the front-end to error if it sees any SYCL_EXTERNAL functions and -fno-sycl-rdc is passed. That is here:  https://github.com/intel/llvm/pull/7595/commits/a7e0464510abe527147877baf421173d9949fc91

I agree changing default code split is kinda weird. So I can implement what I suggested before, support device=auto but error if sycl-post-link needs to do split=none and -fno-sycl-rdc is passed.

What do you think?

Reply:
Looks correct to me. I'm working on the device split changes. Only question is the warning.

Reply:
I think we should make it illegal to declare any function with `SYCL_EXTERNAL` when using the `-fno-sycl-rdc` flag.  Thus, I think we should include a statement like this in the definition of that flag:

> When compiling with this flag, it is invalid for the source file to declare any device function with the `SYCL_EXTERNAL` macro.

And the compiler will emit an error any time a device function has SYCL_EXTERNAL.

The reason I think this is good is because user's might do something like this:

* Source file "one.cpp" defines a function "foo" with SYCL_EXTERNAL and calls it from a kernel in that same TU.  This source file is compiled with "-fno-sycl-rdc".
* Source file "two.cpp" declares (not defines) "foo" with SYCL_EXTERNAL and calls it from a kernel in that TU.  This source file is *not* compiled with "-fno-sycl-rdc".

This scenario won't work, and my proposal will guarantee that an error is emitted.

Remember that the only useful reason to declare a function with SYCL_EXTERNAL is to call it from a different TU.  We don't need to optimize for the weird case where an application defines a function with SYCL_EXTERNAL and then only calls that function from the same TU.

Reply:
Good point, thanks. I will make sure SYCL_EXTERNAL with -fno-sycl-rdc errors regardless of weather it's a declaration or definition.

Reply:
Thanks, @gmlueck 

> We don't need to optimize for the weird case where an application defines a function with SYCL_EXTERNAL and then only calls that function from the same TU.

+1 here